transfer/custom.go

```diff
@@ -0,0 +1,376 @@
+	}
+	// Capture stderr to trace
+	tracer := &traceWriter{}
+	tracer.processName = filepath.Base(a.path)
+	cmd.Stderr = tracer
+	err = cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to start custom transfer command %q remote: %v", a.path, err)
+	}
+	// Set up buffered reader/writer since we operate on lines
+	ctx := &customAdapterWorkerContext{workerNum, cmd, outp, bufio.NewReader(outp), inp, tracer}
+
+	// send initiate message
+	initReq := NewCustomAdapterInitRequest(a.getOperationName(), a.concurrent, a.originalConcurrency)
+	resp, err := a.exchangeMessage(ctx, initReq)
+	if err != nil {
+		a.abortWorkerProcess(ctx)
+		return nil, err
+	}
+	if resp.Error != nil {
+		a.abortWorkerProcess(ctx)
+		return nil, fmt.Errorf("Error initializing custom adapter %q worker %d: %v", a.name, workerNum, resp.Error.Error())
+	}
+
+	tracerx.Printf("xfer: started custom adapter process %q for worker %d OK", a.path, workerNum)
+
+	// Save this process context and use in future callbacks
+	return ctx, nil
+}
+
+func (a *customAdapter) getOperationName() string {
+	if a.direction == Download {
+		return "download"
+	}
+	return "upload"
+}
+
+// sendMessage sends a JSON message to the custom adapter process
+func (a *customAdapter) sendMessage(ctx *customAdapterWorkerContext, req interface{}) error {
+	b, err := json.Marshal(req)
+	if err != nil {
+		return err
+	}
+	tracerx.Printf("xfer: Custom adapter worker %d sending message: %v", ctx.workerNum, string(b))
+	// Line oriented JSON
+	b = append(b, '\n')
+	_, err = ctx.stdin.Write(b)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *customAdapter) readResponse(ctx *customAdapterWorkerContext) (*customAdapterResponseMessage, error) {
+	line, err := ctx.bufferedOut.ReadString('\n')
+	if err != nil {
+		return nil, err
+	}
+	tracerx.Printf("xfer: Custom adapter worker %d received response: %v", ctx.workerNum, line)
+	resp := &customAdapterResponseMessage{}
+	err = json.Unmarshal([]byte(line), resp)
+	return resp, err
+}
+
+// exchangeMessage sends a message to a process and reads a response if resp != nil
+// Only fatal errors to communicate return an error, errors may be embedded in reply
+func (a *customAdapter) exchangeMessage(ctx *customAdapterWorkerContext, req interface{}) (*customAdapterResponseMessage, error) {
+
+	err := a.sendMessage(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return a.readResponse(ctx)
+}
+
+// shutdownWorkerProcess terminates gracefully a custom adapter process
+// returns an error if it couldn't shut down gracefully (caller may abortWorkerProcess)
+func (a *customAdapter) shutdownWorkerProcess(ctx *customAdapterWorkerContext) error {
+	defer ctx.errTracer.Flush()
+
+	tracerx.Printf("xfer: Shutting down adapter worker %d", ctx.workerNum)
+	termReq := NewCustomAdapterTerminateRequest()
+	err := a.sendMessage(ctx, termReq)
+	if err != nil {
+		return err
+	}
+	ctx.stdin.Close()
+	ctx.stdout.Close()
+	return ctx.cmd.Wait()
```

Comment:
OK, but that's an persistent problem unless we add timeouts to _all_ requests to the adapter, in which case we need to define what is a 'reasonable' time between receiving some kind of response - including a progress update or a transfer completion. Right now the equivalent for HTTP transfers is 30 minutes (the keepalive), since the dial / TLS timeouts only deal with unresponsiveness on the initial connection.

I'll add it to this shutdown sequence specifically but there are definitely other stall paths which can be caused by either servers or now adapters.